wgpu/src/backend/direct.rs

```diff
@@ -912,8 +913,19 @@ impl crate::Context for Context {
                     block_ctx_dump_prefix: None,
                 };
                 let parser = naga::front::spv::Frontend::new(spv.iter().cloned(), &options);
-                let module = parser.parse().unwrap();
-                wgc::pipeline::ShaderModuleSource::Naga(Owned(module))
+                match parser.parse() {
+                    Ok(m) => wgc::pipeline::ShaderModuleSource::Naga(Owned(m)),
+                    Err(cause) => {
+                        self.handle_error(
+                            &device_data.error_sink,
+                            cause,
+                            LABEL,
+                            desc.label,
+                            "Device::create_shader_module",
+                        );
+                        wgc::pipeline::ShaderModuleSource::Dummy(PhantomData)
```

Comment:
Hey! I looked into adding `ShaderModuleSource::Invalid` as you suggested, but realized that this (and my use of `ShaderModuleSource::Dummy`) ends up leading to duplicate errors being reported which seems to be why I ended up with multiple errors on parse failure during `Device.on_uncaptured_error` -- I would get one error from the parse error handled using `handle_error` in `wgpu/src/backend/wgpu_core.rs` then another when handling the `ShaderModuleSource::Dummy` variant in `wgpu-core/src/device/resource.rs` (which previously just resulted in a panic).

So I tried a couple different approaches involving `ShaderModuleSource::Invalid` that didn't feel right, but then I realized why none of the approaches up to this point felt right...because in `wgpu/src/backend/wgpu_core.rs` there is no parsing of `wgsl` and hence no handling of parse errors at that level. This led me to the eventual conclusion that the cleanest way to improve error handling for `glsl` is to add `ShaderModuleSource::Glsl` and `ShaderModuleSource::SpirV` (the latter isn't necessary for me, but I figured i would make the change to both as long as I was working on the glsl variant), return that from `ContextWgpuCore.device_create_shader_module` and do the parsing for `glsl` and `spirv` in the same place it's done for wgsl.

This required adding a couple more variants to `CreateShaderModuleError` specific to `glsl` and `spirv` as well as the updating `naga::front::glsl::Frontend.parse` to return a `Result<_, naga::front::glsl::error::ParseError>` where `naga::front::glsl::error::ParseError` is a type I've introduced to encapsulate the `Vec<naga::front::glsl::error::Error>` that was being returned before (i gave up trying to use a vec in `CreateShaderModuleError::ParsingGlsl` after fighting with trait bounds somewhere).

Let me know if this makes sense!

Reply:
Oh yeah, just in case you don't get pinged automatically by new messages: @nical

Reply:
Since this approach involves larger naga changes I'll deffer to @teoxoy or @jimblandy